tests/protocol/test_goal_state.py

```diff
@@ -57,4 +64,40 @@ def test_update_goal_state_should_save_goal_state(self):
         extensions = vm_settings["extensionGoalStates"]
         self.assertEqual(4, len(extensions), "Expected 4 extensions in the test vmSettings")
         for e in extensions:
-            self.assertEqual(e["settings"][0]["protectedSettings"], "*** REDACTED ***", "The protected settings for {0} were not redacted".format(e["name"]))
\ No newline at end of file
+            self.assertEqual(e["settings"][0]["protectedSettings"], "*** REDACTED ***", "The protected settings for {0} were not redacted".format(e["name"]))
+
+
+    def test_update_extensions_goal_state_should_should_retry_on_resource_gone_error(self, _):
+        """
+        Requests to the hostgaplugin incude the Container ID and the RoleConfigName as headers; when the hostgaplugin returns GONE (HTTP status 410) the agent
+        needs to get a new goal state and retry the request with updated values for the Container ID and RoleConfigName headers.
+        """
+        with mock_wire_protocol(mockwiredata.DATA_FILE_VM_SETTINGS) as protocol:
+            # The mock_wire_protocol mocks azurelinuxagent.common.utils.restutil.http_request, but the GONE status is handled at a lower level, in the internal _http_request.
```

Comment:
If you understand what is going on on this unit test and its mock you get extra points... :)

Reply:
(part of the difficulty of testing this code is in the error handling done in the networking code (restutil, etc)